crates/chain/src/local_chain.rs

```diff
@@ -0,0 +1,184 @@
+
+        // the lower bound of the range to invalidate
+        let invalidate_lb = match agreement_height {
+            Some(height) if height == update_tip => u32::MAX,
+            Some(height) => height + 1,
+            None => 0,
+        };
+
+        // the first block's height to invalidate in the local chain
+        let invalidate_from = self.blocks.range(invalidate_lb..).next().map(|(&h, _)| h);
+
+        // the first block of height to invalidate (if any) should be represented in the update
+        if let Some(first_invalid) = invalidate_from {
+            if !update.contains_key(&first_invalid) {
+                return Err(UpdateError::NotConnected(first_invalid));
+            }
+        }
+
+        let invalidated_heights = invalidate_from
+            .into_iter()
+            .flat_map(|from_height| self.blocks.range(from_height..).map(|(h, _)| h));
+
+        // invalidated heights must all exist in the update
+        let mut missing_heights = Vec::<u32>::new();
+        for invalidated_height in invalidated_heights {
+            if !update.contains_key(invalidated_height) {
+                missing_heights.push(*invalidated_height);
+            }
+        }
+        if !missing_heights.is_empty() {
+            return Err(UpdateError::MissingHeightsInUpdate(missing_heights));
+        }
+
+        let mut changeset = BTreeMap::<u32, BlockHash>::new();
+        for (height, new_hash) in update {
+            let original_hash = self.blocks.get(height);
+            if Some(new_hash) != original_hash {
+                changeset.insert(*height, *new_hash);
+            }
+        }
+        Ok(ChangeSet(changeset))
+    }
+
+    /// Applies the given `changeset`.
+    pub fn apply_changeset(&mut self, mut changeset: ChangeSet) {
+        self.blocks.append(&mut changeset.0)
+    }
+
+    /// Updates [`LocalChain`] with an update [`LocalChain`].
+    ///
+    /// This is equivilant to calling [`determine_changeset`] and [`apply_changeset`] in sequence.
+    ///
+    /// [`determine_changeset`]: Self::determine_changeset
+    /// [`apply_changeset`]: Self::apply_changeset
+    pub fn apply_update(&mut self, update: Self) -> Result<ChangeSet, UpdateError> {
+        let changeset = self.determine_changeset(&update)?;
+        self.apply_changeset(changeset.clone());
+        Ok(changeset)
+    }
+
+    pub fn initial_changeset(&self) -> ChangeSet {
+        ChangeSet(self.blocks.clone())
+    }
+
+    pub fn heights(&self) -> BTreeSet<u32> {
+        self.blocks.keys().cloned().collect()
+    }
+}
+
+/// This is the return value of [`determine_changeset`] and represents changes to [`LocalChain`].
+///
+/// [`determine_changeset`]: LocalChain::determine_changeset
+#[derive(Debug, Default, Clone, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(crate = "serde_crate")
+)]
+pub struct ChangeSet(pub BTreeMap<u32, BlockHash>);
```

Comment:
any reason not to just use a `BtreeMap` directly here?

Reply:
I didn't want the caller to directly modify the `ChangeSet`.

Reply:
Already changed to:
```rust
pub type ChangeSet = BTreeMap<u32, Option<BlockHash>>;
```